### PR TITLE
Builder: fix "COPY --from" to non-existing directory on Windows

### LINKED ISSUE #4349: Windows multi-stage builds broken in 19.03.1
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: `FBE0859A-7747-4099-9C4D-E97F0513F17F/20190731191818`

### Expected behavior

Able to build multi-stage Windows images

### Actual behavior

Fails with:

```
Step 6/7 : COPY --from=builder ["C:/install/test.txt", "C:/install/"]
failed to copy files: mkdir \\?: The filename, directory name, or volume label syntax is incorrect.
```

### Information

- Multi-stage builds was working fine on previous stable version: 18.09.0
- Windows Version: Windows 10 Enterprise 1903, build 18362.239
- Docker Desktop Version: 2.1.0.0

### Steps to reproduce the behavior

Can be reproduced with this dockerfile:

```Dockerfile
FROM mcr.microsoft.com/windows/servercore:1903 as builder

RUN MKDIR C:\install
RUN ECHO 'test' > C:\install\test.txt
RUN DIR C:\install

FROM mcr.microsoft.com/windows/servercore:1903

# FAILS!
COPY --from=builder ["C:/install/test.txt", "C:/install/"]

RUN DIR C:\install
```

Build with: `docker image build --tag test .`

### PR DESCRIPTION
**Pull request by thaJeztah**
fixes docker/for-win#4349
built on top of https://github.com/moby/moby/pull/39698

This fixes a regression introduced in 6d87f19142f86b8fee75af721b583a306202f228 (https://github.com/moby/moby/pull/38599),
causing `COPY --from` to fail if the target directory does not exist:

```Dockerfile
FROM mcr.microsoft.com/windows/servercore:ltsc2019 as s1
RUN echo "Hello World" > /hello

FROM mcr.microsoft.com/windows/servercore:ltsc2019
COPY --from=s1 /hello /hello/another/world
```

Would produce an error:

```
Step 4/4 : COPY --from=s1 /hello /hello/another/world
failed to copy files: mkdir \\?: The filename, directory name, or volume label syntax is incorrect.
```

The cause for this was that Go's `os.MkdirAll()` does not support/detect volume GUID paths
(`\\?\Volume{dae8d3ac-b9a1-11e9-88eb-e8554b2ba1db}\hello\another}`), and as a result
attempted to create the volume as a directory (`\\?`), causing it to fail.

This patch replaces `os.MkdirAll()` with our own `system.MkdirAll()` function, which
is capable of detecting GUID volumes.

**- How to verify it**

To verify (on Linux; don't know the equivalent on Windows, LOL):

```
make TEST_FILTER=TestBuildMultiStageCopy test-integration

...
INFO: Testing against a local daemon
=== RUN   TestBuildMultiStageCopy
=== RUN   TestBuildMultiStageCopy/copy_to_root
=== RUN   TestBuildMultiStageCopy/copy_to_newdir
=== RUN   TestBuildMultiStageCopy/copy_to_newdir_nested
=== RUN   TestBuildMultiStageCopy/copy_to_existingdir
=== RUN   TestBuildMultiStageCopy/copy_to_newsubdir
--- PASS: TestBuildMultiStageCopy (11.32s)
    --- PASS: TestBuildMultiStageCopy/copy_to_root (5.76s)
    --- PASS: TestBuildMultiStageCopy/copy_to_newdir (1.33s)
    --- PASS: TestBuildMultiStageCopy/copy_to_newdir_nested (1.41s)
    --- PASS: TestBuildMultiStageCopy/copy_to_existingdir (1.34s)
    --- PASS: TestBuildMultiStageCopy/copy_to_newsubdir (1.48s)
PASS
...
```

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @ddebroy @StefanScherer PTAL

**Review by StefanScherer — APPROVED** [review]
LGTM !

**Comment by thaJeztah** [comment]
ah, dang; guess I need to flip a slash

```
11:12:59 === RUN   TestBuildMultiStageCopy
11:12:59 --- FAIL: TestBuildMultiStageCopy (0.00s)
11:12:59     build_test.go:148: assertion failed: error is not nil: open testdata/Dockerfile.TestBuildMultiStageCopy: The system cannot find the path specified.

**Comment by thaJeztah** [comment]
Fixed (hopefully)

**Comment by StefanScherer** [comment]
still looks good on Windows

**Comment by thaJeztah** [comment]
Still doesn't work on Windows 😞 

```
09:38:32 === RUN   TestBuildMultiStageCopy
09:38:32 --- FAIL: TestBuildMultiStageCopy (0.00s)
09:38:32     build_test.go:149: assertion failed: error is not nil: open testdata\Dockerfile.TestBuildMultiStageCopy: The system cannot find the path specified.
```

**Comment by thaJeztah** [comment]
Getting close

```
12:19:18 --- FAIL: TestBuildMultiStageCopy (0.00s)
12:19:18     build_test.go:151: assertion failed: error is not nil: open C:\gopath\src\github.com\docker\docker\testdata\Dockerfile.TestBuildMultiStageCopy: The system cannot find the path specified.
```

File: integration/build/build_test.go [hunk 1]
```diff
@@ -136,6 +137,56 @@ func buildContainerIdsFilter(buildOutput io.Reader) (filters.Args, error) {
...
 
+// TestBuildMultiStageCopy verifies that copying between stages works correctly
+// regression test for docker/for-win#4349, ENGCORE-935, where creating the target
+// directory failed on Windows, because `os.MkdirAll()` was called with a volume
+// GUID path (\\?\Volume{dae8d3ac-b9a1-11e9-88eb-e8554b2ba1db}\newdir\hello}),
+// which currently isn't supported by Golang.
+func TestBuildMultiStageCopy(t *testing.T) {
+	ctx := context.Background()
+
+	f, err := filepath.Abs("testdata/Dockerfile." + t.Name())
```
Comment (StefanScherer) [inline integration/build/build_test.go]: Maybe inline it as in other tests?

Comment (thaJeztah) [reply integration/build/build_test.go]: That didn't work; looks like on Windows the CWD is different, and it sees `C:\gopath\src\github.com\docker\docker\` as current directory, not `C:\gopath\src\github.com\docker\docker\integration\build\`.

I can work around that, but looking _why_

CI is running the integration tests from here;
https://github.com/moby/moby/blob/4fb5e9e696fb50d65966a4b6b3c7906046d7c6a4/hack/make.ps1#L338-L354

Which looks quite similar to the Linux equivalent;

https://github.com/moby/moby/blob/4fb5e9e696fb50d65966a4b6b3c7906046d7c6a4/hack/make/.integration-test-helpers#L44-L48 


I wonder if `System.Diagnostics.ProcessStartInfo` inherits the working directory, or if it needs to be set explicitly 🤔

Comment (thaJeztah) [reply integration/build/build_test.go]: Looks like that may be it: it has a `ProcessStartInfo.WorkingDirectory` property

From the documentation; https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.processstartinfo.workingdirectory?view=netframework-4.8

> When `UseShellExecute` is `true`, the fully qualified name of the directory that contains
> the process to be started. When the `UseShellExecute` property is `false`, the working
> directory for the process to be started. The default is an empty string (`""`).

Comment (thaJeztah) [reply integration/build/build_test.go]: Opened https://github.com/moby/moby/pull/39698 to fix that

I'll rebase this PR on top of that one to get the fix in this branch

**Comment by thaJeztah** [comment]
hmf test is failing

```
13:42:07 === RUN   TestBuildMultiStageCopy
13:42:07 === RUN   TestBuildMultiStageCopy/copy_to_root
13:42:10 === RUN   TestBuildMultiStageCopy/copy_to_newdir
13:42:12 === RUN   TestBuildMultiStageCopy/copy_to_newdir_nested
13:42:14 === RUN   TestBuildMultiStageCopy/copy_to_existingdir
13:42:16 === RUN   TestBuildMultiStageCopy/copy_to_newsubdir
13:42:18 --- FAIL: TestBuildMultiStageCopy (11.00s)
13:42:18     --- FAIL: TestBuildMultiStageCopy/copy_to_root (2.86s)
13:42:18         build_test.go:183: assertion failed: error is not nil: Error: No such image: testbuildmultistagecopy/copy_to_root
13:42:18     --- FAIL: TestBuildMultiStageCopy/copy_to_newdir (2.16s)
13:42:18         build_test.go:183: assertion failed: error is not nil: Error: No such image: testbuildmultistagecopy/copy_to_newdir
13:42:18     --- FAIL: TestBuildMultiStageCopy/copy_to_newdir_nested (2.07s)
13:42:18         build_test.go:183: assertion failed: error is not nil: Error: No such image: testbuildmultistagecopy/copy_to_newdir_nested
13:42:18     --- FAIL: TestBuildMultiStageCopy/copy_to_existingdir (1.98s)
13:42:18         build_test.go:183: assertion failed: error is not nil: Error: No such image: testbuildmultistagecopy/copy_to_existingdir
13:42:18     --- FAIL: TestBuildMultiStageCopy/copy_to_newsubdir (1.92s)
13:42:18         build_test.go:183: assertion failed: error is not nil: Error: No such image: testbuildmultistagecopy/copy_to_newsubdir

**Comment by thaJeztah** [comment]
no output on those failures. perhaps it's also a race condition (build not yet completed?)

**Comment by thaJeztah** [comment]
Whoop! Fixed the test; now works both on Windows and Linux

- RS1 https://jenkins.dockerproject.org/job/Docker-PRs-WoW-RS1/26337/console
- RS5 https://jenkins.dockerproject.org/job/Docker-PRs-WoW-RS5-Process/3402/console
- Janky https://jenkins.dockerproject.org/job/Docker-PRs/55327/console


```
14:25:17 === RUN   TestBuildMultiStageCopy
14:25:17 === RUN   TestBuildMultiStageCopy/copy_to_root
14:25:28 === RUN   TestBuildMultiStageCopy/copy_to_newdir
14:25:29 === RUN   TestBuildMultiStageCopy/copy_to_newdir_nested
14:25:30 === RUN   TestBuildMultiStageCopy/copy_to_existingdir
14:25:33 === RUN   TestBuildMultiStageCopy/copy_to_newsubdir
14:25:35 --- PASS: TestBuildMultiStageCopy (18.27s)
14:25:35     --- PASS: TestBuildMultiStageCopy/copy_to_root (11.18s)
14:25:35     --- PASS: TestBuildMultiStageCopy/copy_to_newdir (0.95s)
14:25:35     --- PASS: TestBuildMultiStageCopy/copy_to_newdir_nested (1.71s)
14:25:35     --- PASS: TestBuildMultiStageCopy/copy_to_existingdir (2.04s)
14:25:35     --- PASS: TestBuildMultiStageCopy/copy_to_newsubdir (2.39s)
```

**Review by tonistiigi — APPROVED** [review]

**Review by ddebroy — APPROVED** [review]
LGTM with one clarification.

**Comment by thaJeztah** [comment]
> with one clarification.

what's the clarification? did yo forget to post that?

**Comment by thaJeztah** [comment]
Looks like Docker Hub made a whoopsie

```
[2019-08-08T16:32:20.347Z] FAIL: docker_cli_pull_test.go:118: DockerHubPullSuite.TestPullFromCentralRegistryImplicitRefParts
[2019-08-08T16:32:20.347Z] 
[2019-08-08T16:32:20.347Z] docker_cli_pull_test.go:145:
[2019-08-08T16:32:20.347Z]     pullFromV2("hello-world")
[2019-08-08T16:32:20.347Z] docker_hub_pull_suite_test.go:73:
[2019-08-08T16:32:20.347Z]     c.Assert(err, checker.IsNil, check.Commentf("%q failed with errors: %s, %v", strings.Join(arg, " "), out, err))
[2019-08-08T16:32:20.347Z] ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc0009c8ce0), Stderr:[]uint8(nil)} ("exit status 1")
[2019-08-08T16:32:20.347Z] ... "hello-world" failed with errors: Using default tag: latest
[2019-08-08T16:32:20.347Z] latest: Pulling from library/hello-world
[2019-08-08T16:32:20.347Z] Get https://registry-1.docker.io/v2/library/hello-world/manifests/sha256:92c7f9c92844bbbb5d0a101b22f7c2a7949e40f8ea90c8b3bc396879d95e899a: EOF
[2019-08-08T16:32:20.347Z] , exit status 1
```

**Comment by kolyshkin** [comment]
#39698 is merged, can you please rebase?

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by ddebroy** [comment]
oops about the note above. I meant to say for Windows `MkdirAllAndChownNew` and `mkdirAs` are just wrappers around `mkdirall`. So the if condition at https://github.com/docker/ee-engine/blob/master/builder/dockerfile/copy.go#L559 is probably not very relevant for Windows. But in the Linux paths, `mkdirAs` does a lot more work.

**Pull request merged**